File: Hashset/leetcode_18.py
# Problem Link ====>>https://leetcode.com/problems/4sum/description/

# ===================>>>✔️ Solution 1 : HashSet
def fourSum(nums, target):
    seen = set()
    ans_set = set()
    n = len(nums)

    for i in range(n):
        for j in range(i + 1, n):
            for k in range(j + 1, n):
                lastNumber = target - (nums[i] + nums[j] + nums[k])

                if lastNumber in seen:
                    ans_set.add(
                        frozenset((nums[i], nums[j], nums[k], lastNumber)))

        seen.add(nums[i])

    ans = []
    for el in ans_set:
        ans.append(list(el))

    return ans


if __name__ == "__main__":
    nums = [1, 0, -1, 0, -2, 2]
    target = 0

    quadruplets = fourSum(nums, target)

    for quadruplet in quadruplets:
        for num in quadruplet:
            print(num, end=" ")
        print()

File: Hashset/leetcode_18.cpp

// Problem Link ====>>https://leetcode.com/problems/4sum/description/

#include <iostream>
#include <vector>
#include <unordered_set>
#include <set>
#include <algorithm>
using namespace std;

class Solution
{
public:
    vector<vector<int>> fourSum(vector<int> &nums, int target)
    {
        unordered_set<int> seen;
        set<multiset<int>> ans_set;
        int n = size(nums);

        for (int i = 0; i < n; i++)
        {
            for (int j = i + 1; j < n; j++)
            {
                for (int k = j + 1; k < n; k++)
                {
                    int lastNumber = target - (nums[i] + nums[j] + nums[k]);

                    if (seen.count(lastNumber))
                    {
                        ans_set.emplace(multiset<int>{nums[i], nums[j], nums[k], lastNumber});
                    }
                }
            }
            seen.insert(nums[i]);
        }
        vector<vector<int>> ans(size(ans_set));
        for_each(begin(ans_set), end(ans_set), [&, i(0)](auto &el) mutable
                 { ans[i++] = vector<int>(begin(el), end(el)); });
        return ans;
    }
};

int main()
{
    vector<int> nums = {1, 0, -1, 0, -2, 2};
    int target = 0;

    Solution solution;
    vector<vector<int>> quadruplets = solution.fourSum(nums, target);

    for (const vector<int> &quadruplet : quadruplets)
    {
        for (const int &num : quadruplet)
        {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}